drv/lpc55-spi-server/src/main.rs

```diff
@@ -8,24 +8,329 @@
+// the real time requirements of feeding the Tx FIFO when the SP clocks out
+// the response message.
+//
+// ROT_IRQ is expected to be an edge triggered interrupt on the SP.
+// ROT_IRQ is de-asserted only after all Tx bytes have drained.
+// That gives a bit more information when looking at logic analyzer traces.
+
+#[derive(Copy, Clone, PartialEq)]
+#[repr(u8)]
+enum RxState {
+    Idle = 0,       // Idle, waiting for message
+    Header = 1,     // Expecting message length LSB, MSB, and MsgType
+    Payload = 2,    // receiving message payload
+    Dispatch = 3,   // Receive complete, message processing is pending.
+    Responding = 4, // Tx frame is queued. Terminal state until next SOT.
+    Invalid = 5,    // bad message or normally ignoring incoming data until SOT.
+    Error = 0xff,   // Errors, Terminal state until next SOT.
+}
+
+impl From<u8> for RxState {
+    fn from(rxstate: u8) -> Self {
+        match rxstate {
+            0 => RxState::Idle,
+            1 => RxState::Header,
+            2 => RxState::Payload,
+            3 => RxState::Dispatch,
+            4 => RxState::Responding,
+            5 => RxState::Invalid,
+            _ => RxState::Error,
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+#[repr(u8)]
+enum TxState {
+    Idle = 0,    // Idle, waiting for message to be queued.
+    Queued = 1,  // Message queued, waiting for start of Tx frame.
+    Writing = 2, // Message is transferring from local buffer to FIFO.
+    Finish = 3,  // Final bytes are being transmitted from FIFO to SPI bus.
+    // Last byte has been clocked out to SPI bus, return to Idle.
+    Error = 0xff, // Error during transmit.
+}
+
+impl From<u8> for TxState {
+    fn from(txstate: u8) -> Self {
+        match txstate {
+            0 => TxState::Idle,
+            1 => TxState::Queued,
+            2 => TxState::Writing,
+            3 => TxState::Finish,
+            _ => TxState::Error,
+        }
+    }
+}
+
 #[derive(Copy, Clone, PartialEq)]
 enum Trace {
-    IRQ,
-    Tx(u8),
-    Rx(u8),
+    AlreadyEnqueuedMsg,
+    BadMsgLength,
+    CannotSendRxFragError,
+    CollectPayload,
+    AssertRotIrq,
+    DeassertRotIrq,
+    Enqueued,
+    EnqueuedMsg(usize),
+    EnqueueFail,
+    EnqueueTooBig,
+    GetThePayload(usize),
+    HeaderCollect,
+    HeaderComplete,
+    HeaderOnly,
+    Initialized,
+    InvalidRxState(RxState),
+    Irq(bool, bool, bool),
+    Line,
+    SSAsserted,
+    SSDeasserted,
+    Loop(u16, u16, TxState, RxState),
+    LStat(bool, bool, bool, bool, bool, bool, bool, u8, u8),
+    PayloadComplete,
+    ResetRxToIdle,
+    Responding,
+    RespondToEcho,
+    RespondToStatus,
+    RespondToUnknown,
+    RxFragment,
+    Rx(RxState, u8, usize, bool, bool),
+    RxState(RxState),
+    Tx(TxState, u8, usize),
+    StartTx(bool, bool),
     None,
 }
 
-ringbuf!(Trace, 64, Trace::None);
+ringbuf!(Trace, 128, Trace::None);
+
+struct TxContext<'a> {
```

Comment:
`TxContext` and `RxContext` seem like a great place to use the [typestate pattern](http://cliffle.com/blog/rust-typestate/).

For example, in `TxContext`, you could use a richer `TxState` that only includes what's used in each phase of the state machine:
```rust
enum TxState {
    /// Idle, waiting for message to be queued.
    Idle,
    /// Message queued, waiting for start of Tx frame.
    Queued { end: u32 },
    /// Message is transferring from local buffer to FIFO.
    Writing { count: u32, end: u32 },
    /// Final bytes are being transmitted from FIFO to SPI bus.
    Finish,
    /// Error during transmit.
    Error,
}
```

Reply:
done